### PR TITLE
Wip/streamlined writing

### PR DESCRIPTION
**Pull request by rrcomtech**

### CODE REVIEW

File: src/bin/playground.cpp [hunk 1]
```diff
@@ -154,37 +158,29 @@ void write_dict_segment_to_disk(const std::shared_ptr<DictionarySegment<int>> se
...
                           *segment->attribute_vector());
-  chunk_file.close();
 }
 
 std::vector<uint32_t> generate_segment_offset_ends(const std::shared_ptr<Chunk> chunk) {
   const auto segment_count = chunk->column_count();
   auto segment_offset_ends = std::vector<uint32_t>(segment_count);
 
-
+  auto offset_end = static_cast<uint32_t>(4 + 4 * segment_count);
```
Comment (SanJSp) [inline src/bin/playground.cpp:180]: why first 4 ? Refactor into variable name
refactor second 4 to byte_size_multiplyier

File: src/bin/playground.cpp [hunk 2]
```diff
@@ -154,37 +158,29 @@ void write_dict_segment_to_disk(const std::shared_ptr<DictionarySegment<int>> se
...
   for (auto segment_index = size_t{0}; segment_index < segment_count; ++segment_index) {
     // 4 Byte Dictionary Size + 4 Byte Attribute Vector Size + 4 Compressed Vector Type ID
-    auto offset_end = uint32_t{12};
-    if (segment_index == 0){
-      offset_end = 12 + 4 + 4 * segment_count;
-    }
+    offset_end += uint32_t{12};
```
Comment (SanJSp) [inline src/bin/playground.cpp:183]: 12 refactor into speaking variable
segment_header_size f.e.

File: src/bin/playground.cpp [hunk 3]
```diff
@@ -305,13 +293,12 @@ std::shared_ptr<Chunk> map_chunk_from_disk(const uint32_t chunk_offset_end) {
   }
   std::cout << std::endl;
 
-  const auto header_offset = chunk_offset_end / 4 + 1 + COLUMN_COUNT;
-
+  const auto header_offset = chunk_offset_end / 4;
 
   for (auto segment_index = size_t{0}; segment_index < COLUMN_COUNT; ++segment_index) {
-    auto segment_offset_end = uint32_t{0};
+    auto segment_offset_end = uint32_t{24 * 4};
```
Comment (SanJSp) [inline src/bin/playground.cpp:299]: why 24 * 4? should be segment_offset_size * byte_size_multiplier (and somehow add 1) - where does the 1 come from?

File: src/bin/playground.cpp [hunk 4]
```diff
@@ -411,6 +362,57 @@ file_header read_file_header(std::string filename) {
...
+  auto offset = uint32_t{sizeof(file_header)};
+  for (auto chunk_ind = uint32_t{0}; chunk_ind < chunks.size(); ++chunk_ind) {
+    const auto segment_offset_ends = generate_segment_offset_ends(chunks[chunk_ind]);
+    offset += segment_offset_ends.back();
+
+    chunk_segment_offset_ends[chunk_ind] = segment_offset_ends;
+    chunk_offset_ends[chunk_ind] = offset;
+  }
+  // Fill all offset fields, that are not used with 0s.
+  for (auto rest_chunk_ind = chunks.size(); rest_chunk_ind < CHUNK_COUNT; ++rest_chunk_ind) {
```
Comment (SanJSp) [inline src/bin/playground.cpp:393]: same as above, index instead of ind

File: src/bin/playground.cpp [hunk 5]
```diff
@@ -411,6 +362,57 @@ file_header read_file_header(std::string filename) {
...
+
+  file_header fh;
+  fh.storage_format_version_id = 2;
+  fh.chunk_count = static_cast<uint16_t>(chunks.size());
+  fh.chunk_ids = chunk_ids;
+  fh.chunk_offset_ends = chunk_offset_ends;
+
+  export_value<file_header>(FILESTREAM, fh);
+
+  for (auto chunk_ind = uint32_t{0}; chunk_ind < chunks.size(); ++chunk_ind) {
```
Comment (SanJSp) [inline src/bin/playground.cpp:408]: ind -> index

File: src/bin/playground.cpp [hunk 6]
```diff
@@ -305,13 +293,12 @@ std::shared_ptr<Chunk> map_chunk_from_disk(const uint32_t chunk_offset_end) {
   }
   std::cout << std::endl;
 
-  const auto header_offset = chunk_offset_end / 4 + 1 + COLUMN_COUNT;
-
+  const auto header_offset = chunk_offset_end / 4;
```
Comment (SanJSp) [inline src/bin/playground.cpp:296]: divide by byte_factor

File: src/bin/playground.cpp (hunk 1 above)
Comment (rrcomtech) [reply src/bin/playground.cpp:179]: Done.

File: src/bin/playground.cpp (hunk 3 above)
Comment (rrcomtech) [reply src/bin/playground.cpp:298]: Done.

File: src/bin/playground.cpp [hunk 7]
```diff
@@ -411,6 +362,57 @@ file_header read_file_header(std::string filename) {
...
+
+  auto chunk_segment_offset_ends = std::vector<std::vector<uint32_t>>(CHUNK_COUNT);
+  auto chunk_offset_ends = std::array<uint32_t, CHUNK_COUNT>();
+  auto chunk_ids = std::array<uint32_t, CHUNK_COUNT>();
+
+  // Chunk Offsets generally start right after the file header.
+  // i.e., an offset of 0 is equivalent to the 102nd byte in the file.
+  // Segment Offsets generally start right after the end of the last chunk.
+  auto offset = uint32_t{sizeof(file_header)};
+  for (auto chunk_ind = uint32_t{0}; chunk_ind < chunks.size(); ++chunk_ind) {
```
[the remark this answers is not in the document]
Comment (rrcomtech) [reply src/bin/playground.cpp:379]: Done.

File: src/bin/playground.cpp (hunk 4 above)
Comment (rrcomtech) [reply src/bin/playground.cpp:387]: Done.

File: src/bin/playground.cpp [hunk 8]
```diff
@@ -411,6 +362,57 @@ file_header read_file_header(std::string filename) {
...
+    chunk_segment_offset_ends[chunk_ind] = segment_offset_ends;
+    chunk_offset_ends[chunk_ind] = offset;
+  }
+  // Fill all offset fields, that are not used with 0s.
+  for (auto rest_chunk_ind = chunks.size(); rest_chunk_ind < CHUNK_COUNT; ++rest_chunk_ind) {
+    chunk_offset_ends[rest_chunk_ind] = uint32_t{0};
+  }
+
+  // TODO(everyone): Find, how to get the actual chunk id.
+  for (auto index = uint32_t{0}; index < CHUNK_COUNT; ++index) chunk_ids[index] = index;
```
[the remark this answers is not in the document]
Comment (rrcomtech) [reply src/bin/playground.cpp:392]: Done.

File: src/bin/playground.cpp (hunk 5 above)
Comment (rrcomtech) [reply src/bin/playground.cpp:402]: Done.

File: src/bin/playground.cpp (hunk 6 above)
Comment (rrcomtech) [reply src/bin/playground.cpp:295]: Done.

**Pull request merged**